src/charm.py

```diff
@@ -642,580 +90,250 @@ def k8s_auth(self) -> None:
-        """
-        api = self._networking_v1_api()
-        ingresses = api.list_namespaced_ingress(namespace=self._namespace)
-        if ingress_name in [x.metadata.name for x in ingresses.items]:
-            api.delete_namespaced_ingress(ingress_name, self._namespace)
-            LOGGER.info(
-                "Ingress deleted in namespace %s with name %s",
-                self._namespace,
-                ingress_name,
-            )
+            options: Configuration options for the ingress. If not provided, no resources will be
+                created but the cleanup will still run.
 
-    def _delete_unused_resources(self) -> None:
-        """Delete unused services and ingresses."""
-        svc_names = [conf_or_rel._service_name for conf_or_rel in self._all_config_or_relations]
-        self._delete_unused_services(svc_names)
-        svc_hostnames = [
-            conf_or_rel._service_hostname for conf_or_rel in self._all_config_or_relations
-        ]
-        all_additional_hostnames = [
-            conf_or_rel._additional_hostnames for conf_or_rel in self._all_config_or_relations
-        ]
-        for additional_hostname in all_additional_hostnames:
-            svc_hostnames.extend(additional_hostname)
-        self._delete_unused_ingresses(svc_hostnames)
-
-    def _on_config_changed(self, _: HookEvent) -> None:
+        Returns:
+            The created or modified ingress resource, None if no ingress resource was created or
+                modified.
+        """
+        endpoints_controller = self._endpoints_controller
+        endpoint_slice_controller = self._endpoint_slice_controller
+        service_controller = self._service_controller
+        ingress_controller = self._ingress_controller
+        namespace = options.service_namespace if options is not None else self.model.name
+        endpoints = None
+        endpoint_slice = None
+        service = None
+        ingress = None
+        define_resource = functools.partial(self._define_resource, namespace=namespace)
+        cleanup_resources = functools.partial(self._cleanup_resources, namespace=namespace)
+        if options is not None and options.use_endpoint_slice:
+            endpoints = define_resource(controller=endpoints_controller, options=options)
+            endpoint_slice = define_resource(controller=endpoint_slice_controller, options=options)
+        if options is not None:
+            service = define_resource(controller=service_controller, options=options)
+            ingress = define_resource(controller=ingress_controller, options=options)
+        cleanup_resources(controller=endpoints_controller, exclude=endpoints)
+        cleanup_resources(controller=endpoint_slice_controller, exclude=endpoint_slice)
+        cleanup_resources(controller=service_controller, exclude=service)
+        cleanup_resources(controller=ingress_controller, exclude=ingress)
+        return ingress
+
+    def _update_ingress(self) -> None:
```

Comment:
1. Check preconditions -> error out if there is a problem
2. If there is no ingress relation, clean up everything and return
3. if there is a relation, take the data and convert it to IngressOptions
4. Create the ingress resources based on IngressOptions
5. cleanup any unnecessary ingress (e.g., created by the previous relation) that weren't just created